gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/oneconvergence_service_node_driver.py

```diff
@@ -1026,9 +1168,148 @@ def _generate_pool_member_template(self, context,
+            context._plugin_context, {'consumer_ptg_id': [new_consumer[
+                'id']]})[0]
+        # REVISIT(VK) Can this be a issue?
+        stack_name = ("stack_" + context.instance['name'] +
+                      context.current_node['name'] +
+                      new_cons_sc_instance['id'][:8] +
+                      context.current_node['id'][:8])
+        # Update template
+        self.update_firewall_template(context, stack_template,
+                                      ptg_to_not_configure=new_consumer['id'])
+        stack = heatclient.create(stack_name, stack_template, {})
+        self._wait_for_stack_operation_complete(
+            heatclient, stack["stack"]["id"], "create")
+        self._insert_node_instance_stack_in_db(
+            context.plugin_session, context.current_node['id'],
+            new_cons_sc_instance['id'], stack['stack']['id'])
+
+    @log.log
+    def get_firewall_stack_id(self, context, cons_ptgs=None):
+        if not cons_ptgs:
+            sc_instances = context._sc_plugin.get_servicechain_instances(
+                context._plugin_context, {'provider_ptg_id': [context.provider[
+                    'id']]})
+        else:
+            for cons_ptg in cons_ptgs:
+                sc_instances = \
+                    context._sc_plugin.get_servicechain_instances(
+                        context._plugin_context,
+                        {'consumer_ptg_id': [cons_ptg]})
+                for sc_instance in sc_instances:
+                    stacks = self._get_node_instance_stacks(
+                        context.plugin_session, context.current_node['id'],
+                        sc_instance['id'])
+                    if stacks:
+                        return stacks, sc_instance
+            return None, None
+        for sc_instance in sc_instances:
+            stacks = self._get_node_instance_stacks(
+                context.plugin_session, context.current_node['id'],
+                sc_instance['id'])
+            if stacks:
+                return stacks, sc_instance
+
+    @log.log
     def delete(self, context):
+        _exist = None
         try:
-            super(OneConvergenceServiceNodeDriver, self).delete(context)
+            _exist, cons_ptgs, provider_unset = \
+                    self.check_for_existing_service(context)
+            service_type = context.current_profile["service_type"]
+            if service_type in [pconst.FIREWALL, pconst.LOADBALANCER]:
+                # if provider_unset:
+                #     super(OneConvergenceServiceNodeDriver, self).delete(
+                #         context)
+                if (_exist and service_type == pconst.FIREWALL and
+                      not context.is_consumer_external and not provider_unset):
+                    self.update_firewall(context, cons_ptgs)
+                else:
+                    super(OneConvergenceServiceNodeDriver, self).delete(
```

Comment:
In case of delete, I think we have an issue for both LB and Firewall. In case of LB, stack should not be deleted if some other consumer exists. And for both firewall and LB, we should also delete the old entry in DB and add a new entry for this stack against another service chain ID that would remain after this consumer is gone.